### PR TITLE
Only load data tables when necessary

### LINKED ISSUE #2315: Don't load $.dataTable on all pages
### Description
The dataTable library  that is currently pulled in for all users is only inside only one template - templates/type/work/editions_dataTable yet the CSS and JS is loaded for all pages that use it. We pull this off the critical path into a conditional loading module. This will reduce the initial JS we ship improving site performance  as well as isolate this code to only the pages that need it.

### Expectation
Site will work just as before. Hopefully a little faster on pages without the editions table

### Details

- **Logged in (Y/N)?** 
- **Browser type/version?** 
- **Operating system?**

### PR DESCRIPTION
**Pull request by jdlrobson**
### Description

Going forward we will use dynamic imports for different pages rather than relying on inline scripts. While this will result in additional requests on some pages, it will do so with asynchronous unblocking JavaScript and reduce the amount of initial JS loaded on every other page.

Close #2315

### Technical
The inline script from openlibrary/templates/type/work/editions_datatable.html
is moved into a new entry point openlibrary/plugins/openlibrary/js/edition-tables.js
the code is exactly the same, with the only changes being styling so it confirms to
our eslint standard.

The index.js entry point will now load this file CONDITIONALLY if an
loaded where it is not needed.

To support dynamic importing a babel plugin is added.
"@babel/plugin-syntax-dynamic-import adds support for the import
function

The babel-eslint parser is needed to avoid eslint being unable to parse this page due to the unexpected import keyboard being used at the bottom of the file.

### Testing
This code is used on https://openlibrary.org/works/OL53924W/The_complete_works_of_Mark_Twain and controls the arrows in the table.
You'll want to confirm that this code is not loaded on any page which doesn't have that table.

### Evidence
You should notice no change, e.g. https://openlibrary.org/works/OL53924W/The_complete_works_of_Mark_Twain matches what you see locally.

### CODE REVIEW

**Comment by cdrini** [comment]
Overall looks good. One question: Instead of using the dynamic import, (which, as you said, requires a JS file the loads another JS file), what about putting script import in the template? i.e. `<script src="./editions-table/index.js`"></script>` in the editions template. Thoughts?

**Comment by jdlrobson** [comment]
> Overall looks good. One question: Instead of using the dynamic import, (which, as you said, requires a JS file the loads another JS file), what about putting script import in the template? i.e. `<script src="./editions-table/index.js`"></script>` in the editions template. Thoughts?

The problem with that is there's potential for having multiple script tags across multiple templates and the developer will need to be aware of where these are and when to take extra caution in test - using one entry point removes this ambiguity.

The code will also load unconditionally. While that doesn't apply here, I think this is going to be more common. In future we'll want to conditionally load jquery ui, my plan is to only load it when something is interacted with - so for example, when you click a list, we will pull down the library so when you add to list the popup code is loaded.

https://github.com/internetarchive/openlibrary/pull/2305/commits/68da2c9918f485bef836f4ac31c069355856a32a shows how that would work.

We can't do this kind of thing using the script tag approach and I think it's important to be consistent.

File: openlibrary/plugins/openlibrary/js/index.js [hunk 1]
```diff
@@ -101,4 +99,9 @@ jQuery(function () {
     initAnalytics($);
     init($);
     initShowPasswords($);
+    // conditionally load functionality based on what's in the page
+    if (document.getElementById('editions')) {
+        import(/* webpackChunkName: "editions-table" */ './editions-table')
+            .then((module) => module.initEditionsTable() );
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/index.js]: Can drop the `()`
```suggestion
            .then(module => module.initEditionsTable());
```

File: webpack.config.js [hunk 2]
```diff
@@ -68,6 +68,9 @@ module.exports = {
     output: {
         // Specify the destination of all build products.
         path: distDir,
+        // base path for build products when referenced from production
+        // (see https://webpack.js.org/guides/code-splitting/)
```
Comment (cdrini) [inline webpack.config.js]: That link doesn't actually mention this parameter :P Use this one: https://webpack.js.org/guides/public-path/ (or no link is also ok)

File: package.json [hunk 3]
```diff
@@ -15,9 +15,13 @@
     "test": "npm run check-bundles && stylelint --syntax less static/css/ && npm run lint:js && npm run test:unit"
   },
   "bundlesize": [
+    {
+      "path": "static/build/edition-tables.js",
```
Comment (cdrini) [inline package.json]:
```suggestion
      "path": "static/build/editions-table.js",
```

**Comment by cdrini** [comment]
> `import()` calls use promises internally. If you use `import()` with older browsers, remember to shim `Promise` using a polyfill such as `es6-promise` or `promise-polyfill`.

Is this an issue for us?

**Comment by cdrini** [comment]
Tested:
- ✅  editions-table.js does not load on homepage
- ✅ editions-table.js DOES load on work page
- ❌ `id=editions` exists only on works template
  - also in https://github.com/internetarchive/openlibrary/blob/f6dcab1fb9237853bbc2c25cce23f829146a58f6/openlibrary/templates/type/list/editions.html
- ❌  IE11 loades editions-table.js on works page
  - fails with `Promise undefined` error

File: openlibrary/plugins/openlibrary/js/index.js (hunk 1 above)
Comment (jdlrobson) [reply openlibrary/plugins/openlibrary/js/index.js]: Done

> > `import()` calls use promises internally. If you use `import()` with older browsers, remember to shim `Promise` using a polyfill such as `es6-promise` or `promise-polyfill`.
> 
> Is this an issue for us? IE11 doesn't support promises.

We support IE11 so yes this matters and yes this does break IE11. Excellent catch! `promise-polyfill` is the most lightweight of these solutions so I'll go with that.

File: webpack.config.js (hunk 2 above)
Comment (jdlrobson) [reply webpack.config.js]: done

File: package.json (hunk 3 above)
Comment (jdlrobson) [reply package.json]: Done.

**Comment by jdlrobson** [comment]
> > `import()` calls use promises internally. If you use `import()` with older browsers, remember to shim `Promise` using a polyfill such as `es6-promise` or `promise-polyfill`.
> 
> Is this an issue for us? IE11 doesn't support promises.

We support IE11 so yes this matters and yes this does break IE11. Excellent catch! `promise-polyfill` is the most lightweight of these solutions so I'll go with that.

**Comment by jdlrobson** [comment]
> * ❌ `id=editions` exists only on works template
>   
>   * also in https://github.com/internetarchive/openlibrary/blob/f6dcab1fb9237853bbc2c25cce23f829146a58f6/openlibrary/templates/type/list/editions.html

Can you elaborate on this line. What's the problem here? Are there other editions table this is supposed to run on?

**Comment by cdrini** [comment]
I can't tell if that template is in use. I was checking if `id=editions` was being used anywhere for NOT an editions datatable.

**Comment by jdlrobson** [comment]
This now works on IE 11:


Screenshot 2019-08-29 at 1 10 09 PM

**Comment by jdlrobson** [comment]
I suspect https://github.com/internetarchive/openlibrary/blob/f6dcab1fb9237853bbc2c25cce23f829146a58f6/openlibrary/templates/type/list/editions.html is dead code.

https://openlibrary.org/people/jdlrobson/lists/OL126798L/editions redirects to another page.

**Comment by jdlrobson** [comment]
I suspect https://github.com/internetarchive/openlibrary/blob/f6dcab1fb9237853bbc2c25cce23f829146a58f6/openlibrary/templates/type/list/editions.html is dead code.

https://openlibrary.org/people/jdlrobson/lists/OL126798L/editions redirects to another page.

If I look at the bottom of https://openlibrary.org/people/jdlrobson/lists I see a similar (updated version of that same table).

Usually type/list/editions.html would correspond to the URI /list/editions but that throws a 404.

Opened up https://github.com/internetarchive/openlibrary/pull/2348.
If for some reason it is accessible I don't see any problem with running a jquery table enhancement plugin on a table.

**Comment by jdlrobson** [comment]
Rebased.
Since we're not sure if we need to support http://localhost:8080/people/openlibrary/lists/OL1L/A_list/editions I've added a follow up that changes the ID selector of that table (and adds a class for styling)

**Comment by cdrini** [comment]
Sorry for the delay, @jdlrobson ; besides the fact that I've been in the middle of a house change this week, there's been something bugging me about this change which I couldn't articulate until ~now.

I think my issue with this change is that it's not explicit when/how the `editions-table` js is used. We have a "component" here, `editions-table`, split across 3 files: the `.html`, the `.js` and the `.css`. The relationship between the latter two is simple and explicit because the CSS is loaded from the JS file. The relationship between the former two is not explicit. You have to check the `index.js` file, see that it relies on an `id=editions` and then look for files that use this ID. That flow doesn't reflect the true state of the code, which is that it's used _only_ by `editions_datatable.html`. Relying on `id` also makes this solution brittle because it's easy for someone to accidentally use that ID on another page and get unexpected outcomes.

Solutions? Adding a `data-dynamic-import="./editions-table"` to the `#editions` table. This makes it explicit, but avoids setting a precedent of importing by looking for an ID which is brittle. Note I don't think this is **the** solution, but I think it's better than the current one, and will also be easier to update to another system than the current solution.

Long term solution might be to change our file structure to group our templates/js by components (as opposed to by file type which it is currently), but I think that's out of the scope of this PR.

Thoughts?

**Comment by jdlrobson** [comment]
My goal is to separate JS from the HTML into one place.

I don't think what I am doing is not explicit.  To me this is just what progressive enhancement is. You run JS based on what's in the page. If the id is being used in multiple places where it's not desired then it's a poorly named id and a poorly organised set of templates (which we have).

In future we will have templates shared on the client and the server side similar to how the JS and CSS are linked, but getting all the JS is the same place is my main focus right now with my limited time.

I think the solution here is the correct one. We can rename the id if you are worried about existing templates e.g. progressively-enhanced-editions-table

**Comment by cdrini** [comment]
> If the id is being used in multiple places where it's not desired then it's a poorly named id and a poorly organised set of templates (which we have).

I don't think the solution to this is to pretend we don't have a "poorly organised set of templates"; the solution is too build something _orthogonal_ to our current templates, and merge them once they are no longer poorly organized. The data attribute solution would allow for this better than the current `id` solution.

Could you address my comments about the fact that this is a component? And that using ID as a binding factor is brittle? And the proposed `data-` solution? I don't think it violates any of the principles you've outlined.

**Comment by jdlrobson** [comment]
To be it's honest I am really struggling to understand the objection here to using ids. Identifiers have been used since the dawn of the web by JS to identify elements for enhancement. document.getElementById enjoys good browser support and performance for this reason. It's an abbreviation for identifier and that's what we're doing.. identifying an element.

Data attributes were not designed to be used in this way, they are for communicating data between the server and client.  Even the spec says this:

"Custom data attributes are intended to store custom data private to the page or application, for which there are no more appropriate attributes or elements."

An id is an appropriate attribute here. 
https://html.spec.whatwg.org/multipage/dom.html#custom-data-attribute

My goal remains clear - to get all Javascript out of html templating while making sure not to bloat the overall JS code so we have a clearer separation of our code.

**Comment by cdrini** [comment]
I feel like we're talking past each other :P Did you read my questions in my last comment? Do you understand what I'm trying to say when I say `editions-table` is a component? Do you agree or disagree? Is structuring our frontend by components still a goal?

The issue with IDs is that they are global identifiers. They have the same problems that global variables have in code. This isn't a huge issue if you have 1, maybe 2 .html files (which was their use case in the "dawn of the web" era), but it falls apart when you have hundreds of templates that can import one another. There's no way to scope and you end up with global namespace collisions (case in point: `id=editions` appearing multiple times). But that's not the point :P My point is that you're separating code that's closely coupled: the html/js of this component.

The _core_ of my issue is that this _is_ a component whose parts, which are (rightfully) tightly coupled, appear loosely coupled.

**Comment by cdrini** [comment]
> The core of my issue is that this is a component whose parts, which are (rightfully) tightly coupled, appear loosely coupled.

Also, this sets a precedent for how we do things which, because IDs are brittle (being global identifiers), would be difficult to change. If we wanted to change this pattern, we'd have to check all e.g. `id=editions` and see if they are actually using the JS.

**Comment by jdlrobson** [comment]
I don't believe we are talking past each other, I just don't feel you are understanding what I am trying to achieve and that's frustrating.

> If we wanted to change this pattern, we'd have to check all e.g. id=editions and see if they are 
> actually using the JS.

This just seems wrong to me. An ids should be unique across all pages and when they are the same it's because they come from the same HTML. We should not have multiple things with id editions and to have templates which share the same ID but contain different things semantically is an anti-pattern. This is something the current codebase is doing wrong IMO and we can fix as part of this. No checking should be necessary. That's the problem right now. I've suggested one approach to this - new identifiers.

> Did you read my questions in my last comment? Do you understand what I'm trying to say when I say editions-table is a component? Do you agree or disagree? Is structuring our frontend by components still a goal?

I really don't see how this is related, but okay. It depends what you are defining as a component. From my perspective, it's an HTML table of data that is being progressively enhanced with JavaScript. In current form there's nothing componenty about it as there is no component framework that defines what a component is.  I think a component should be reusable and should be a collection of HTML5 elements assembled in some particular way. At best it's a table component.

> My point is that you're separating code that's closely coupled: the html/js of this component.

CSS is also tightly coupled and I'd hope you'd agree that now it's a little easier to manage since the CSS is in stylesheets organized by page. My only goal right now is to get all the JavaScript of the site into one entry point in a similar way, separated from the HTML and make it more manageable in a similar way to how I made the CSS more manageable. Right now, changes to JS can break templates that run JavaScript. Making templates more manageable is not a goal of mine right now.  I just don't have the time to commit to doing that with my 3 hours a week.

**Comment by cdrini** [comment]
This is why I feel we're talking past each other :P If you ever find yourself thinking "I really don't see how this is related," chances are I haven't explained myself well, so please say so! I wouldn't of have brought it up if I didn't think it was related :)

I apologize this is dragging on; I'm trying to be sensitive of your time, but I think this is important. I'm trying to find a solution which would require as few changes as possible to your code (which I think the data attribute solution does).

> ids should be unique across all pages

Regardless of whether they should be or not, they aren't. So using IDs attaches your code to a 100+ file global namespace. That's a maintainability nightmare, which I think we should avoid. (That's why I'm also wary of changing this ID to something else, because who knows if there isn't some jQuery somewhere that's searching for this element by ID and making changes to it, that will break if you change the ID? It'll make this PR _significantly_ harder to test.)

Ok, now the conversation I was trying to have 3 comments ago! A component is usually defined by a model, a view (in HTML/CSS), and a controller (in JS) (e.g. vue: https://vuejs.org/v2/guide/single-file-components.html). Your definition is dropping the `model` part, which is important. The fact that this can display a list of editions (as opposed to some other list) is what makes it different from a generic table component. Here we have the model=editions, view=HTML/CSS, controller=JS file. So I think this is a component. The only reason it's not reusable atm is because it makes use of a global namespace (through IDs/etc). I thought your main goal was to separate things into design components (see below). Is this correct?

```
components/
  search-bar/
    search-bar.html
    search-bar.css
    search-bar.js
  editions-table/
    editions-table.html
    editions-table.css
    editions-table.js
```

With respect to the CSS changes, I haven't followed them too closely, but I'm sure they've improved on what was there :) I think putting the JS in their own file is an improvement, but I think the way we're linking them is a step away from the component structure which I thought we were aiming for.

**Comment by jdlrobson** [comment]
> Regardless of whether they should be or not, they aren't. So using IDs attaches your code to a 100+ file global namespace. That's a maintainability nightmare, which I think we should avoid. (That's why I'm also wary of changing this ID to something else, because who knows if there isn't some jQuery somewhere that's searching for this element by ID and making changes to it, that will break if you change the ID? It'll make this PR significantly harder to test.)

I really don't understand the problem you are describing. There is nothing rule with accessing IDs from JavaScript. That's not a maintainability nightmare... If we really can't agree on that point, I don't know how to proceed on this patch or any of the others I have planned.

>  component is usually defined by a model, a view (in HTML/CSS), and a controller (in JS) (e.g. vue: https://vuejs.org/v2/guide/single-file-components.html).

Yes but we are not using Vue.js right now. We have a typical jQuery site of HTML with JS running on top of it. I don't see how any of this relates right now to what I'm doing. If we have a single JS file of the existing code and can see how the code is organized

> I thought your main goal was to separate things into design components (see below). Is this correct?
Aspirational goal, yes, but not my current goal. Again, my goal right now is to get JS organized in a single place and out of templates with the minimum amount of work required. Without doing this, quite frankly I'm lost and I don't see how I can do any of this.

**Comment by mekarpeles** [comment]
this all lgtm

**Comment by cdrini** [comment]
We're having a call to resolve the remaining issue (probably) today :+1:

**Comment by mekarpeles** [comment]
Happy to test tomorrow, I have a few branches in testing on dev already :)

Thanks @jdlrobson + @cdrini for your hard work on this

**Comment by cdrini** [comment]
Tested:
- ✔ Loads editions-table.js on http://localhost:8080/works/OL53924W/The_complete_works_of_Mark_Twain
- ✔ Doesn't load editions-table.js on http://localhost:8080/
- ✔ Editions pagination works
- ✔ Check "Return eBook" button works

**Pull request merged**